# src/database/connMongo.py
from motor.motor_asyncio import AsyncIOMotorClient
from fastapi import HTTPException, status
from typing import Optional
from src.database.modelConfig.configMongo import MongoParamBuilder 

class MongoConnectionManager:
    client: Optional[AsyncIOMotorClient] = None
    DB_NAME: str = "" 

    @classmethod
    async def connect(cls):
        try:
            builder = MongoParamBuilder()
            config = builder.config
        except Exception as e:
            print(f"ERRO DE CONFIGURA√á√ÉO DO AMBIENTE MONGODB: {e}")
            raise RuntimeError(f"Falha ao carregar as vari√°veis de ambiente do MongoDB: {e}") 
        
        MONGO_URI = config.mongo_uri
        cls.DB_NAME = config.mongo_db_name # Usa o nome do DB validado pelo Pydantic

        try:
            cls.client = AsyncIOMotorClient(MONGO_URI)
            await cls.client.admin.command('ping') 
            print(f"MongoConnectionManager: Cliente MongoDB Atlas ativo (DB: {cls.DB_NAME}).")
        except Exception as e:
            cls.client = None
            print(f"ERRO FATAL ao conectar ao MongoDB Atlas: {e}")
            raise Exception(f"Falha na conex√£o com MongoDB: {e}") 

    @classmethod
    async def close(cls):
        if cls.client:
            cls.client.close()
            cls.client = None
            
    @classmethod
    def get_client(cls) -> AsyncIOMotorClient:
         if not cls.client:
             raise HTTPException(
                status_code=status.HTTP_503_SERVICE_UNAVAILABLE, 
                detail="Cliente MongoDB n√£o est√° ativo. Falha no Startup."
            )
         return cls.client
    

# if __name__ == "__main__":
#     import asyncio
    
#     async def test_connection():
#         print("\n--- üß™ Teste de Conex√£o Direta ao MongoDB Atlas ---")
#         try:
#             # Tenta conectar
#             await MongoConnectionManager.connect()
            
            
            
#             print("\n‚úîÔ∏è SUCESSO! A conex√£o com o MongoDB Atlas foi estabelecida com sucesso.")
            
#         except Exception as e:
#             # Se a conex√£o falhar, o erro detalhado ser√° impresso pelo m√©todo connect()
#             print("\n‚ùå FALHA: N√£o foi poss√≠vel estabelecer a conex√£o com o MongoDB Atlas.")
#             # O detalhe exato do erro deve ter sido impresso acima
            
#         finally:
#             # Garante que a conex√£o seja fechada ap√≥s o teste
#             await MongoConnectionManager.close()
#             print("Conex√£o fechada.")
            
#     # Executa a fun√ß√£o ass√≠ncrona
#     asyncio.run(test_connection())